applications/sentence-transformers-ft/finetune.py

```diff
@@ -0,0 +1,103 @@
+    losses,
+    evaluation,
+    models,
+)
+from torch import nn
+import pathlib
+
+
+# TODO: refactor this to separate finetune from modal code.
+# maybe initialize the dataset inside the function?
+# also maybe make the hyperparameters as parameters of the finetune function
+def finetune(
+    model_id: str,
+    save_path: pathlib.Path,
+    epochs: int = 10,
+    dataset_fraction: int = 1,
+    use_dense_layer: bool = True,
+    dense_out_features: int = 200,
+):
+    """
+    Finetune a sentence transformer on the quora pairs dataset. Evaluates model performance before/after training
+
+    :returns: evaluation accuracy post training
+    :rtype: float
+    Inspired by: https://github.com/UKPLab/sentence-transformers/blob/657da5fe23fe36058cbd9657aec6c7688260dd1f/examples/training/quora_duplicate_questions/training_MultipleNegativesRankingLoss.py
+    """
+
+    # Quora pairs dataset: https://huggingface.co/datasets/quora
+    DATASET_ID = "quora"
+    dataset = load_dataset(DATASET_ID, split="train")
+    # Quora pairs dataset only contains a "train" split in huggingface, so we will manually split it into train and test
+    train_test_split = dataset.train_test_split(test_size=0.1, seed=42)
+    train_dataset = train_test_split["train"]
+    test_dataset = train_test_split["test"]
+
+    if use_dense_layer:
+        embedding_model = SentenceTransformer(model_id)
+        dense_model = models.Dense(
+            in_features=embedding_model.get_sentence_embedding_dimension(),
+            out_features=dense_out_features,
+            activation_function=nn.Tanh(),
+        )
+        model = SentenceTransformer(modules=[embedding_model, dense_model])
+    else:
+        model = SentenceTransformer(model_id)
+
+    train_examples = []
+    # TODO: can make this more pythonic later by removing dataset_fraction
+    for i in range(train_dataset.num_rows // dataset_fraction):
+        text0 = train_dataset[i]["questions"]["text"][0]
+        text1 = train_dataset[i]["questions"]["text"][1]
+        is_duplicate = int(train_dataset[i]["is_duplicate"])
+        train_examples.append(InputExample(texts=[text0, text1], label=is_duplicate))
+    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=64)
+    train_loss = losses.OnlineContrastiveLoss(model)
+
+    test_examples = []
+    # TODO: can make this more pythonic later by removing dataset_fraction
+    for i in range(test_dataset.num_rows // dataset_fraction):
+        text0 = test_dataset[i]["questions"]["text"][0]
+        text1 = test_dataset[i]["questions"]["text"][1]
+        is_duplicate = int(test_dataset[i]["is_duplicate"])
+        test_examples.append(InputExample(texts=[text0, text1], label=is_duplicate))
+    evaluator = evaluation.BinaryClassificationEvaluator.from_input_examples(
+        test_examples,
+    )
+
+    # evaluator.name is used for how the file name is saved
+    evaluator.csv_file = "binary_classification_evaluation_pre_train" + "_results.csv"
```

Comment:
should we be making edits to the attribute?

Reply:
I think it's ok to modify this attribute. the `self.csv_file` variable in the `BinaryClassificationEvaluator` class is only used once when naming the csv file that the evaluator writes results to https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/BinaryClassificationEvaluator.py#L97
```
        if output_path is not None and self.write_csv:
            csv_path = os.path.join(output_path, self.csv_file)
            if not os.path.isfile(csv_path):
                with open(csv_path, newline='', mode="w", encoding="utf-8") as f:
                    writer = csv.writer(f)
                    writer.writerow(self.csv_headers)
                    writer.writerow(file_output_data)
            else:
                with open(csv_path, newline='', mode="a", encoding="utf-8") as f:
                    writer = csv.writer(f)
                    writer.writerow(file_output_data)
```

In this case we edit it to manually save the pre-training and post-training evaluations into different CSV files so we can see both of them and compare. Also, after modifying the pre-training csv_file name, I revert it to it's default name (note we don't set a name attribute so that part is empty)
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/BinaryClassificationEvaluator.py#L51
```self.csv_file = "binary_classification_evaluation" + ("_"+name if name else '') + "_results.csv"```